### DISCUSSION TITLE
Making ARTIS database more portable 🥡 and lightweight 🪶

### DISCUSSION THREAD

**Discussion by theamarks** [Seafood-Globalization-Lab/artis-model#26] [Ideas]
## Background 

Our current workflow involves researchers downloading trade data (~12GB) and consumption data (~7GB) and then setting up a local `pgAdmin` database to query data and pull relevant portions into RStudio for analysis. Or often a quick short cut is to filter the ARTIS database `.csv`'s to a custom data subset, because reading the .csv sizes often run laptops out of memory. Both avenues are cumbersome and create version consistency uncertainty that will only grow more challenging as the volume and complexity of the data continue to grow. 

## Proposed Solution 

We are exploring two potential database formats -- `DuckDB` or `SQLite` -- as alternatives to our current file-based approach. The aim is to package the ARTIS database in a format that offers efficient data access on a local computer, simplified version management, and seamless integration with our existing analytical tools.

## Benefits of Transitioning

1. **Improved Performance**: Both `DuckDB` and `SQLite` provide faster querying and better handling of large datasets compared to traditional `CSV` files. `DuckDB`, in particular, is optimized for analytical queries.
2. **Enhanced Data Integrity**: Using a single database file for each model version reduces the risk of mismatched file versions and simplifies the process of distributing updates to users.
3. **Seamless Integration**: 
   - **DuckDB**: Offers strong performance for analytical workloads and integrates with R and Python using `duckplyr`, allowing users to query data directly into local memory and work with familiar `dplyr` syntax.
   - **SQLite**: Provides a lightweight, widely supported option that works well for smaller datasets and offers robust support across multiple platforms and tools - `Rsqlite` package
4. **Flexible Access**: Both `DuckDB` and `SQLite` can be used as portable, local databases that users can query directly without complex setup or external dependencies. 

## Next Steps
1. **Pilot Testing**: Convert the ARTIS database into both DuckDB and SQLite formats. We will evaluate performance, ease of use, and compatibility with existing workflows.
2. **Feedback and Discussion**: We invite feedback from the community on the following:
   - Preferred database format (DuckDB vs. SQLite) for our use case.
   - Experiences with `duckplyr` or `dbplyr` when working with these databases.
   - Any anticipated challenges or concerns with this transition.

3. **Documentation and Training**: Based on the results of the pilot tests, we will develop documentation and training materials to help users transition to the new database format smoothly.

## How to Get Involved
We encourage you to share your thoughts, experiences, and any concerns you may have about transitioning the ARTIS database to `DuckDB` or `SQLite`. Your input will help us make an informed decision and ensure that the new database format meets the needs of our diverse user community.

**Comment by theamarks** [comment]
| **Aspect**               | **DuckDB**                                   | **SQLite**                              |
|--------------------------|----------------------------------------------|-----------------------------------------|
| **Performance**          | :white_check_mark: High-performance for analytical queries and large datasets. Efficient for complex operations and aggregations.         | :yellow_circle: Performs well for simple queries and smaller datasets. Not as efficient for complex or large analytical tasks.  |
| **File Size Handling**   | :white_check_mark: Efficiently handles large datasets, including direct access to Parquet and CSV files.  | :yellow_circle: Can handle moderate-sized datasets effectively, but performance may degrade with very large files.         |
| **Integration**          | :white_check_mark: Strong integration with R and Python using `duckplyr`.<br>:white_check_mark: Directly query Parquet files. | :white_check_mark: Broad integration with many tools.<br>:white_check_mark: Works well with `dbplyr` in R. |
| **Data Integrity**       | :white_check_mark: Manages complex data types and large-scale data operations without data loss.           | :white_check_mark: Reliable for basic data integrity. Suitable for small to moderate datasets. |
| **Flexibility**          | :white_check_mark: Optimized for analytical workloads and supports complex operations, even with large datasets.                     | :yellow_circle: Ideal for simple CRUD operations and lightweight applications.         |
| **Platform Compatibility**| :white_check_mark: Cross-platform support (Windows, macOS, Linux).<br>:yellow_circle: May have some limitations on Windows for direct file queries. | :white_check_mark: Excellent cross-platform support.<br>:white_check_mark: Very stable across all platforms. |
| **Storage Efficiency**   | :white_check_mark: Supports efficient compression and columnar storage formats (e.g., Parquet).            | :yellow_circle: Limited to row-based storage, which can be less efficient for large data. |
| **Concurrency**          | :white_check_mark: Suitable for low concurrency, single-user environments. Scales well for large, analytical queries. | :white_check_mark: Suitable for low concurrency use cases, single-user applications, and embedded systems. |
| **Use Cases**            | :white_check_mark: Ideal for analytical tasks, data exploration, and working with large external files.    | :white_check_mark: Best for small, single-user applications, lightweight data management, and embedded systems. |
| **Deployment**           | :white_check_mark: Portable as a single file.<br>:white_check_mark: No server needed.<br>:white_check_mark: Can work with multiple file types directly. | :white_check_mark: Portable as a single file.<br>:white_check_mark: Easy to set up and use. |
| **Metadata and Version Control** | :white_check_mark: Supports metadata storage within the DuckDB file. Can easily maintain version history by adding metadata tables or fields within the database. <br>:white_check_mark: Compatible with file-based version control systems like Git for database files. | :white_check_mark: Supports metadata storage in custom tables. Easy to track database versions using file-based systems like Git. <br>:white_check_mark: Limited support for complex versioning within the database itself. |
| **Maximum Database Size**| :white_check_mark: Can handle large datasets up to petabytes. Efficient querying allows working with datasets much larger than available RAM (e.g., 16 GB RAM can handle hundreds of GBs). | :white_check_mark: Can handle large datasets up to 140 TB. Practical performance is best with datasets up to several GBs using efficient queries. |

**Comment by theamarks** [comment]
@Anurag19101996 wanted to point you to this discussion, I've been keeping track of the ARTIS `duckdb` here. Feel free to add comments, questions, or just document your findings here.

**Comment by theamarks** [comment]
# Planned Contents for ARTIS `DuckDB` 🦆 

@Anurag19101996 

## Background
Moving forward we will ingest new data and rerun the ARTIS model ~1x / year. There are 2 "modes" for the model depending on which data is input into the model. These "modes" are: 
  - `FAO` (Food and Agriculture Organization of the United Nations) and 
  -  `SAU` (Sea Around Us). 

Both ARTIS model "modes" produce similarly named (if not the same) output files. Ultimately we will have 2 `.duckdb` databases to distinguish the two ARTIS database "modes". This will ensure that end users are querying and analyzing a set of files that all correspond with the exact same model run. 

## Outline for each `.duckdb` file

### ARTIS_FAO_[model-version]_[run-date].duckdb

#### Tables:

- trade 
  - primary database table
  - sometimes also referred to as "snet"
  - [KNB](https://knb.ecoinformatics.org/view/doi%3A10.5063%2FF1CZ35N7) has this data in the `data/trade/` directory divided up by `HS` version and `year` (e.g. `artis_midpoint_HS02_2016.csv`)
- consumption 
  - primary database table
  - KNB has this data in the `data/consumption/` directory divided up by `HS` version and `year` (e.g. `consumption_midpoint_HS12_2012.csv`)
- code_max_resolved 
  - attribute table  
  - KNB `data/attribute tables/code_max_resolved.csv`
- countries
  - attribute table
  - KNB `data/attribute tables/countries.csv`
- products
  - attribute table
  - KNB `data/attribute tables/products.csv`
- sciname
  - attribute table
  - KNB `data/attribute tables/sciname.csv`
- standardized_sau_pop
  - attribute table
  - KNB `data/attribute tables/standardized_fao_pop.csv` 
  
### ARTIS_SAU_[model-version]_[run-date].duckdb

#### Tables:

- trade 
  - primary database table
  - sometimes also referred to as "snet"
  - [Google Drive folder ](https://drive.google.com/drive/u/3/folders/1mtYFEuBtWWgs76qL8rs_8DNVfZBKnTP2) has this data aggregated into a single file `snet_midpoint_all_hs_all_years.parquet`
- consumption 
  - primary database table
  - Google Drive folder has this data aggregated into a single file `2024_09_12_SAU_consumption_midpoint.parquet`
- baci
  - attribute table
  - Google Drive `files-for-duckdb/sql_database/baci.csv`
- code_max_resolved
  - attribute table  
  - Google Drive `files-for-duckdb/sql_database/code_max_resolved.csv`
- countries
  - attribute table
  - Google Drive `files-for-duckdb/sql_database/countries.csv`
- products
  - attribute table
  - Google Drive `files-for-duckdb/sql_database/products.csv`
- sciname
  - attribute table
  - Google Drive `files-for-duckdb/sql_database/sciname.csv`

**Reply by Anurag19101996** [reply]
Started working on this, will upload one test file here:
https://github.com/Seafood-Globalization-Lab/artis-duckdb-metadata

**Comment by theamarks** [comment]
## Compare metadata schemas

*This table is a template created by ChatGPT, @Anurag19101996 please modify as you see fit*

| Feature                         | EML (Ecological Metadata Language) | TBD Database Metadata Schema |
|---------------------------------|----------------------------------|-----------------------------|
| **Purpose**                     | Describe ecological datasets     | [Add purpose here]          |
| **Standard Format**              | XML-based                        | [Add format here]           |
| **Versioning Support**           | Yes, supports dataset versioning | [Add versioning details]    |
| **Data Types Supported**         | Tabular, spatial, qualitative   | [Add supported types]       |
| **Schema Extensibility**         | Modular, supports extensions     | [Describe extensibility]    |
| **Controlled Vocabularies**      | Uses formal vocabularies         | [Specify vocab support]     |
| **Integration with Other Systems** | Supports data repositories (e.g., DataONE) | [Describe integrations] |
| **Documentation Depth**          | Highly detailed, includes methods and provenance | [Describe documentation] |
| **Machine Readability**          | Fully machine-readable           | [Specify level]            |
| **Human Readability**            | Somewhat complex, verbose XML    | [Specify level]            |
| **Use Cases**                    | Ecological research, data sharing | [List key use cases]      |
| **Adoption by Research Community** | Widely used in ecology and environmental sciences | [Describe adoption] |

**Reply by theamarks** [reply]
More details in this issue - https://github.com/Seafood-Globalization-Lab/artis-duckdb-metadata/issues/2

**Comment by theamarks** [comment]
## Make it easy to build ARTIS duckdb 

💡 **Function in exploreARTIS that pulls directly from KNB data repository (where we permanently archive ARTIS data) and builds local duckdb in a standardized way.** Makes use of existing infrastructure and makes distributing new ARTIS database as easy as calling a single function. More details here --> https://github.com/Seafood-Globalization-Lab/exploreARTIS/issues/28
